src/hotspot/share/runtime/jniHandles.cpp

```diff
@@ -325,6 +326,40 @@ bool JNIHandles::current_thread_in_native() {
           JavaThread::current()->thread_state() == _thread_in_native);
 }
 
+bool JNIHandles::is_same_object(jobject handle1, jobject handle2) {
+  oop obj1 = resolve_no_keepalive(handle1);
+  oop obj2 = resolve_no_keepalive(handle2);
+
+  bool ret = obj1 == obj2;
+
+  if (EnableValhalla) {
+    if (!ret && obj1 != NULL && obj2 != NULL && obj1->klass() == obj2->klass() && obj1->klass()->is_inline_klass()) {
+      Thread* THREAD = Thread::current();
+        Handle ha(THREAD, obj1);
+        Handle hb(THREAD, obj2);
+        JavaValue result(T_BOOLEAN);
+        JavaCallArguments args;
+        args.push_oop(ha);
+        args.push_oop(hb);
+        methodHandle method(THREAD, Universe::is_substitutable_method());
+        JavaCalls::call(&result, method, &args, THREAD);
+        if (HAS_PENDING_EXCEPTION) {
+          // Something really bad happened because isSubstitutable() should not throw exceptions
+          // If it is an error, just let it propagate
+          // If it is an exception, wrap it into an InternalError
+          if (!PENDING_EXCEPTION->is_a(SystemDictionary::Error_klass())) {
+            Handle e(THREAD, PENDING_EXCEPTION);
+            CLEAR_PENDING_EXCEPTION;
+            THROW_MSG_CAUSE_(vmSymbols::java_lang_InternalError(), "Internal error in substitutability test", e, false);
+          }
+        }
+        ret = result.get_jboolean();
+    }
```

Comment:
indentation issue:  could lines 338 - 356 should be shifted 2 spaces to the left.

Reply:
Fixed

Reply:
The changes look good!  Thanks for making them.